### PR TITLE
Feature homa

### PR DESCRIPTION
**Pull request by danmanj**

### CODE REVIEW

**Review by oserres — COMMENTED** [review]
LGTM

File: protocol_driver_grpc.cc [hunk 1]
```diff
@@ -155,15 +209,17 @@ GrpcInlineServerDriver::~GrpcInlineServerDriver() {}
 absl::Status GrpcInlineServerDriver::Initialize(
     const ProtocolDriverOptions& pd_opts, int* port) {
   std::string netdev_name = pd_opts.netdev_name();
-  auto maybe_ip = IpAddressForDevice(netdev_name);
+  transport_ =
+      GetNamedServerSettingString(pd_opts, "transport", DEFAULT_TRANSPORT);
+  auto maybe_ip = IpAddressForDevice(netdev_name, transport_ == "homa" ? 4 : 0);
```
[the remark this answers is not in the document]
Comment (danmanj) [reply protocol_driver_grpc.cc:214]: The user code does, but kernel support is not merged upstream yet.

File: protocol_driver_grpc.cc [hunk 2]
```diff
@@ -139,12 +184,21 @@ class TrafficService : public Traffic::Service {
...
-    return grpc::Status::OK;
+    handler_set_.WaitForNotification();
+    if (handler_) {
+      auto remaining_work = handler_(&rpc_state);
+      if (remaining_work) {
+        remaining_work();
+      }
+      return grpc::Status::OK;
+    } else {
+      return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
```
[the remark this answers is not in the document]
Comment (danmanj) [reply protocol_driver_grpc.cc:195]: Woops must be a paste error. I'll change it.

File: protocol_driver_homa.cc [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+                                     std::function<void(void)> done_callback) {
+  PendingHomaRpc* new_rpc = new PendingHomaRpc;
+
+  new_rpc->done_callback = done_callback;
+  new_rpc->state = state;
+  new_rpc->serialized_request = "?";  // Homa can't send a 0 byte message :(
+  state->request.AppendToString(&new_rpc->serialized_request);
+
+  ++pending_rpcs_;
+
```
[the remark this answers is not in the document]
Comment (danmanj) [reply protocol_driver_homa.cc:222]: I'd like to not have a fixed limit on message length in the homa implementation.

File: protocol_driver_homa.cc [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    ServerRpcState* rpc_state = new ServerRpcState;
+    rpc_state->request = request;
+    rpc_state->SetFreeStateFunction([=]() {
+      delete rpc_state->request;
+      delete rpc_state;
+    });
+    rpc_state->SetSendResponseFunction([=, &pending_actionlist_threads]() {
+      std::string txbuf = "!";  // Homa can't send a 0 byte message :(
+      rpc_state->response.AppendToString(&txbuf);
+      int64_t error = homa_reply(homa_server_sock_, txbuf.c_str(),
```
[the remark this answers is not in the document]
Comment (danmanj) [reply protocol_driver_homa.cc:269]: Might be nice to check error recovery by letting the kernel complain, and see that user mode does the right thing in response. I don't want to have 2 layers of parameter checking, let the kernel do it.

File: protocol_driver_test.cc [hunk 5]
```diff
@@ -280,6 +280,19 @@ ProtocolDriverOptions GrpcCallbackClientInlineServer() {
   return pdo;
 }
 
+ProtocolDriverOptions HomaOptions() {
```
[the remark this answers is not in the document]
Comment (danmanj) [reply protocol_driver_test.cc:283]: This is parallel to MercuryOptions.

File: protocol_driver_test.cc [hunk 6]
```diff
@@ -280,6 +280,19 @@ ProtocolDriverOptions GrpcCallbackClientInlineServer() {
...
+  ProtocolDriverOptions ret;
+  ret.set_protocol_name("homa");
+  return ret;
+}
+
+ProtocolDriverOptions Homa(ProtocolDriverOptions pdo) {
```
[the remark this answers is not in the document]
Comment (danmanj) [reply protocol_driver_test.cc]: done

File: protocol_driver_grpc.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (danmanj) [reply protocol_driver_grpc.cc]: Ok v6 was merged upstream.

File: docs/getting-started.md [hunk 7]
```diff
@@ -10,7 +10,7 @@ For general information about Distbench, please consult the
 
 ### Dependencies
 
-To build Distbench, you need to have Bazel installed.
+To build Distbench, you need to have Bazel and g++11.3 installed.
```
Comment (oserres) [inline docs/getting-started.md:13]: Due to homa ? We should fix the Kokoro test scripts too; not sure if 11.3 is available there...

File: protocol_driver_homa.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (oserres) [reply protocol_driver_homa.cc:222]: We have the limit on the rx path though... just wanted to have it check earlier vs later as I am not sure how it's handled on rx (message held until a big enough buffer is provided ? truncated ? errcode ?)

**Review by oserres — APPROVED** [review]
LGTM

File: protocol_driver_homa.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (danmanj) [reply protocol_driver_homa.cc:222]: it's truncated actually.

[the remark this answers is not in the document]
Comment (danmanj) [reply protocol_driver_homa.cc:222]: Yes, so Homa has a limit of 1000000 bytes per message as of now. We will need to convince upstream to handle larger messages, and update the API to peek ahead to see how much buffer to allocate. Until then this is mediocre, but fine.

[the remark this answers is not in the document]
Comment (oserres) [reply protocol_driver_homa.cc:222]: Ok. Sounds good.

**Pull request merged**